File: ts-core/tests/retry.test.ts
import { withRetry } from '../src/retry.js';

describe('withRetry', () => {
  it('returns result on success', async () => {
    const result = await withRetry(async () => 'ok');
    expect(result).toBe('ok');
  });

  it('retries on retryable error', async () => {
    let attempts = 0;
    const result = await withRetry(
      async () => {
        attempts++;
        if (attempts < 3) throw { code: 'NETWORK_ERROR', message: 'fail' };
        return 'ok';
      },
      { maxAttempts: 3, baseDelayMs: 1 },
    );
    expect(result).toBe('ok');
    expect(attempts).toBe(3);
  });

  it('throws after max attempts', async () => {
    await expect(
      withRetry(
        async () => {
          throw { code: 'NETWORK_ERROR', message: 'fail' };
        },
        { maxAttempts: 2, baseDelayMs: 1 },
      ),
    ).rejects.toMatchObject({ code: 'NETWORK_ERROR' });
  });

  it('does not retry on non-retryable error', async () => {
    let attempts = 0;
    await expect(
      withRetry(
        async () => {
          attempts++;
          throw { code: 'VALIDATION_ERROR', message: 'fail' };
        },
        { maxAttempts: 3, baseDelayMs: 1 },
      ),
    ).rejects.toMatchObject({ code: 'VALIDATION_ERROR' });
    expect(attempts).toBe(1);
  });

  it('accepts custom retryOn function', async () => {
    let attempts = 0;
    const result = await withRetry(
      async () => {
        attempts++;
        if (attempts < 2) throw new Error('custom');
        return 'ok';
      },
      { maxAttempts: 3, baseDelayMs: 1, retryOn: () => true },
    );
    expect(result).toBe('ok');
    expect(attempts).toBe(2);
  });
});

File: dart_core/test/retry_test.dart
import 'package:telebirr_sdk_core/src/retry.dart';
import 'package:test/test.dart';

void main() {
  group('withRetry', () {
    test('returns result on success', () async {
      final result = await withRetry(() async => 'ok');
      expect(result, 'ok');
    });

    test('retries on retryable error', () async {
      var attempts = 0;
      final result = await withRetry(
        () async {
          attempts++;
          if (attempts < 3) throw Exception('Network error');
          return 'ok';
        },
        config: RetryConfig(
          maxAttempts: 3,
          baseDelay: Duration(milliseconds: 1),
          retryOn: (e) => true,
        ),
      );
      expect(result, 'ok');
      expect(attempts, 3);
    });

    test('throws after max attempts', () async {
      expect(
        () => withRetry(
          () async {
            throw Exception('fail');
          },
          config: RetryConfig(
            maxAttempts: 2,
            baseDelay: Duration(milliseconds: 1),
            retryOn: (e) => true,
          ),
        ),
        throwsA(isA<Exception>()),
      );
    });

    test('does not retry on non-retryable error', () async {
      var attempts = 0;
      expect(
        () => withRetry(
          () async {
            attempts++;
            throw Exception('Validation error');
          },
          config: RetryConfig(
            maxAttempts: 3,
            baseDelay: Duration(milliseconds: 1),
          ),
        ),
        throwsA(isA<Exception>()),
      );
      expect(attempts, 1);
    });
  });
}
